pkg/policy/repository.go

```diff
@@ -351,26 +353,51 @@ nextLabel:
 // This is just a helper function for unit testing.
 // TODO: this should be in a test_helpers.go file or something similar
 // so we can clearly delineate what helpers are for testing.
-func (p *Repository) Add(r api.Rule) (uint64, error) {
+func (p *Repository) Add(r api.Rule, localRuleConsumers []IdentityConsumer) (uint64, map[uint16]struct{}, error) {
 	p.Mutex.Lock()
 	defer p.Mutex.Unlock()
 
 	if err := r.Sanitize(); err != nil {
-		return p.revision, err
+		return p.revision, nil, err
 	}
 
 	newList := make([]*api.Rule, 1)
 	newList[0] = &r
-	return p.AddListLocked(newList), nil
+	rev := p.AddListLocked(newList, localRuleConsumers, NewIdSet(), &sync.WaitGroup{})
+	return rev, map[uint16]struct{}{}, nil
+}
+
+func NewIdSet() *IdSet {
+	return &IdSet{
+		IDs: map[uint16]struct{}{},
+	}
+}
+
+type IdentityConsumer interface {
+	GetID16() uint16
+	GetSecurityIdentity() *identity.Identity
 }
 
 // AddListLocked inserts a rule into the policy repository with the repository already locked
 // Expects that the entire rule list has already been sanitized.
-func (p *Repository) AddListLocked(rules api.Rules) uint64 {
-	newList := make([]*rule, len(rules))
+func (p *Repository) AddListLocked(rules api.Rules, localRuleConsumers []IdentityConsumer, consumersToUpdate *IdSet, policySelectionWG *sync.WaitGroup) uint64 {
+	policySelectionWG.Add(len(localRuleConsumers))
+
+	newList := make(ruleSlice, len(rules))
 	for i := range rules {
-		newList[i] = &rule{Rule: *rules[i]}
+		newRule := &rule{
+			Rule:               *rules[i],
+			localRuleConsumers: map[uint16]*identity.Identity{},
+			processedConsumers: map[uint16]struct{}{},
+		}
+		newList[i] = newRule
+	}
+
+	for _, identityConsumer := range localRuleConsumers {
+		// Spawn goroutine per rule to avoid blocking on matching via API
```

Comment:
Change to "avoid blocking API calls with costly LabelSelector matching operations"